src/librustc_mir/transform/const_prop.rs

```diff
@@ -374,10 +383,59 @@ impl<'a, 'mir, 'tcx> ConstPropagator<'a, 'mir, 'tcx> {
                 Some(ImmTy::from_scalar(mplace.ptr.into(), place_layout).into())
             },
             Rvalue::Repeat(..) |
-            Rvalue::Aggregate(..) |
             Rvalue::NullaryOp(NullOp::Box, _) |
             Rvalue::Discriminant(..) => None,
 
+            Rvalue::Aggregate(ref kind, ref operands) => {
+                match **kind {
+                    AggregateKind::Tuple => {
+                        if operands.len() == 2 {
+                            let one = self.eval_operand(&operands[0], source_info)?;
+                            let two = self.eval_operand(&operands[1], source_info)?;
+
+                            Some(ImmTy {
+                                imm: Immediate::ScalarPair(
+                                    self.opty_to_scalar(one)?,
+                                    self.opty_to_scalar(two)?
+                                ),
+                                layout: place_layout
+                            }.into())
+                        } else if operands.len() == 1 {
+                            let opty = self.eval_operand(&operands[0], source_info)?;
+
+                            Some(ImmTy {
+                                imm: Immediate::Scalar(self.opty_to_scalar(opty)?),
+                                layout: place_layout
+                            }.into())
+                        } else {
+                            let mplace = self.use_ecx(source_info, |this| {
+                                Ok(this.ecx.allocate(place_layout, MemoryKind::Stack))
+                            }).unwrap();
+
+                            for (i, operand) in operands.iter().enumerate() {
+                                let op = self.eval_operand(operand, source_info)?;
+                                let should_continue = self.use_ecx(source_info, |this| {
+                                    if let Ok(imm) = this.ecx.try_read_immediate(op)? {
```

Comment:
There's no need to check for scalars. You can write the `op` directly to the `field` without having to know whether it was an immediate. This way you can create arbitrary tuple combinations like `(5, (6, 7, 8))` and not just `(5, 6, 7, 8)`.